### PR TITLE
Add new lessons and remove old intro

### PR DESCRIPTION
**Pull request by leosoaivan**
- Remove old "Intro to RSpec"
- Add "TDD"
- Add new "Intro to RSpec"

### CODE REVIEW

File: ruby_programming/testing_with_rspec/test_driven_development.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Test-Driven Development
+
+## Introduction
+
+Hopefully, you've been having fun developing in Ruby thus far. Perhaps one thing you may **not** have enjoyed as much, though, is having to run your script on the command line, running through each step manually, only to find that it doesn't return what you expected. And then comes the *even-less-fun* part: You tug at your hair a bit, change a few variables, and run your script again through each step, **hoping** that the results are correct.
+
+Folks, we're here to tell you there *IS* better. Hoping is for that exam you didn't study for and took by the seat of your pants. **Knowing** is test-driven development.
```
Comment (105ron) [inline ruby_programming/testing_with_rspec/test_driven_development.md]: Folks, we're here to tell you there *IS* a better way.

File: ruby_programming/testing_with_rspec/test_driven_development.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+## Why Is It Important?
+
+Even if TDD intuitively seems like good practice, its utility is hotly debated even today. 
+
+No one is saying that tests aren't useful. Testing at [all levels](https://robots.thoughtbot.com/rails-test-types-and-the-testing-pyramid) is crucial. Testing ensures that your application is working as intended. Testing provides a safety net that gives your application room to grow and evolve as painlessly as possible as features are added and dependencies change.
+
+The real question is whether or not tests should *drive* development, or whether or not it's important for them to come *first*. Many developers write tests after they've written their code. There is definitely a longer list of pros than cons for having tests in general—no one will disagree with that—so should we test first?
```
Comment (105ron) [inline ruby_programming/testing_with_rspec/test_driven_development.md]: There is definitely a longer list of pros than cons for having tests in general, no one will disagree with that. So should we test first?

File: ruby_programming/testing_with_rspec/test_driven_development.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+The real question is whether or not tests should *drive* development, or whether or not it's important for them to come *first*. Many developers write tests after they've written their code. There is definitely a longer list of pros than cons for having tests in general—no one will disagree with that—so should we test first? 
+
+As usual, it depends. Still, here are some reasons we think it might be important for you to implement TDD:
+
+1. **Practice**. Many employers look for developers who have experience or interest in testing. Sure, you can write tests later, but will you really? Test first for good measure.
+2. **Effectiveness**. Writing tests before code is like stretching before a workout. It's not required, but your workout won't be as good. Writing tests first ensures that the code that comes after works and is of high quality.
+3. **Efficiency**. You'll save time in the long run by not having to re-run entire applications just to make sure little parts work.
+
+## Assignments
```
Comment (105ron) [inline ruby_programming/testing_with_rspec/test_driven_development.md]: ## Assignment
<div class="lesson-content__panel" markdown="1">
+ add wrapping div with attributes `"lesson-content__panel" markdown="1"`

File: ruby_programming/testing_with_rspec/test_driven_development.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+1. Read up on [9 Benefits of Test-Driven Development](https://www.madetech.com/blog/9-benefits-of-test-driven-development) for additional insight on this practice. Note that these points may be more applicable to testing in general rather than TDD specifically, but they are still good points nonetheless.
+2. For a more rounded take, consider DHH's ["TDD is Dead. Long Live TDD"](http://david.heinemeierhansson.com/2014/tdd-is-dead-long-live-testing.html), which sparked controversial conversations on what TDD is, what it has become, and what it should be.
```
Comment (105ron) [inline ruby_programming/testing_with_rspec/test_driven_development.md]: 2. For a more rounded take, consider DHH's ["TDD is Dead. Long Live TDD"](http://david.heinemeierhansson.com/2014/tdd-is-dead-long-live-testing.html), which sparked controversial conversations on what TDD is, what it has become, and what it should be.
</div>

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
+# Introduction to RSpec
+
+## Introduction
+In the previous lesson, we established the utility of [test-driven development](https://en.wikipedia.org/wiki/Test-driven_development) (TDD) in maintaining your code and sanity. In this lesson, we'll introduce you to your new best friend, the [RSpec](http://rspec.info/) testing framework. RSpec is one of the most popular testing frameworks, having been downloaded over [100 million times](https://rubygems.org/gems/rspec), as of this writing, and having been ported for use in [Rails testing](https://rubygems.org/gems/rspec-rails).
```
Comment (I3uckwheat) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: "as of the time of this writing" maybe?

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+At this point, you may be wondering, *Why RSpec? Surely, there are other frameworks out there*, and you'd be right. There are. In fact, at one point, Ruby came bundled with [Test::Unit]((https://github.com/test-unit/test-unit)) and later [Minitest](https://github.com/seattlerb/minitest) as part of its standard library, the latter of which lives on in Rails. If you tend to be pessimistic (I'm sorry, I meant *realistic*), then the [Wrong](https://github.com/sconover/wrong) testing framework might be your cup of tea. Or perhaps you're hungry and in the mood for something more substantial, in which case a side of [Bacon](https://github.com/chneukirchen/bacon) might be what you need. At the end of the day, it doesn't matter which framework you choose as long as you cultivate your testing skills. RSpec's wider adoption, especially in the Rails community, is certainly reason enough to warrant familiarity with it, but implementing tests should be the end-all and be-all, rather than choosing a particular framework.
+
+But enough proselytizing. Strap your helmet and buckle up; we're going to jump right in.
```
Comment (I3uckwheat) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Maybe link https://www.dictionary.com/browse/proselytizing on the crazy word?

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 5 above)
Comment (mindovermiles262) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Or just omit it.

`RSpec .. having been downloaded over 100 million times.`

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+## Introduction
+In the previous lesson, we established the utility of [test-driven development](https://en.wikipedia.org/wiki/Test-driven_development) (TDD) in maintaining your code and sanity. In this lesson, we'll introduce you to your new best friend, the [RSpec](http://rspec.info/) testing framework. RSpec is one of the most popular testing frameworks, having been downloaded over [100 million times](https://rubygems.org/gems/rspec), as of this writing, and having been ported for use in [Rails testing](https://rubygems.org/gems/rspec-rails).
+
+## Learning Outcomes
+*Look through these now and use them to guide your learning. By the end of this lesson, expect to:*
+
+* Know what RSpec is
+* Know how to install RSpec
+* Understand basic RSpec syntax:
```
Comment (mindovermiles262) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Understand **the** basic RSpec syntax

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+```
+
+As expected, the output will read:
+
+```bash
+  create   .rspec
+  create   spec/spec_helper.rb
+```
+
+Run the tests from your terminal by using the `rspec` command, which will return "No examples found." That really shouldn't surprise you, because we haven't written any tests yet. If you're still shocked...maybe take a short break, or come say hello to us in our [Ruby Gitter](https://gitter.im/TheOdinProject/Ruby) room.
```
Comment (mindovermiles262) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: The ruby room no longer exists - Update link to Back-End room:
`https://gitter.im/TheOdinProject/Back-End`

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+  def add(a,b)
+  end
+end
+```
+
+Finally, we must also tell the spec where the `Calculator` class is being defined. This is easily done with `require`:
+
+```ruby
+#spec/calculator_spec.rb
+require './lib/calculator'
```
Comment (mindovermiles262) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Add an `# => Added this` comment to the `require` line to make it stand out. SEE: Line 188 for example

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+  end
+end
+```
+
+2. Write a test for a new `Calculator` method (`#multiply`, `#subtract`, or `#divide`) using a new `describe` block. Include at least one `it` block with an appropriate expectation clause. Get it to pass, and refactor if necessary.
+3. In the terminal, try running your failing or passing tests with `rspec --format documentation`. What's different?
+4. RSpec reads command line configurations from `.rspec`, one of the two files generated when RSpec is initialized in a project. If you liked the output you got with `--format documentation`, you can use the `.rspec` file to hold that flag. In doing so, you won't have to type it in every time you run your test suite. Open the file in your text editor and, on a new line, add `--format documentation`. For more information on configuring RSpec, see the docs [here](https://relishapp.com/rspec/rspec-core/v/3-7/docs/configuration).
+
+## Additional Resources
+*This section contains helpful links to other content. It isn't required, so consider it supplemental for if you need to dive deeper into something*
```
Comment (mindovermiles262) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: * Solidify these concepts with a [shamelss plug](https://medium.com/@mindovermiles262/getting-started-with-rspec-part-1-9418909f5e53) from another Odin Project contributor

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 11]
```diff
@@ -0,0 +1,236 @@
+# Introduction to RSpec
```
Comment (KevinMulhern) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Can we remove this heading please? this will be generated on the site automatically 🙂

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+
+## Introduction
+In the previous lesson, we established the utility of [test-driven development](https://en.wikipedia.org/wiki/Test-driven_development) (TDD) in maintaining your code and sanity. In this lesson, we'll introduce you to your new best friend, the [RSpec](http://rspec.info/) testing framework. RSpec is one of the most popular testing frameworks, having been downloaded over [100 million times](https://rubygems.org/gems/rspec), as of this writing, and having been ported for use in [Rails testing](https://rubygems.org/gems/rspec-rails).
+
+## Learning Outcomes
```
Comment (KevinMulhern) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Can we change all the h2 headings denoted with `##` to h3's please?  The styling on the main site expects h3's for any top level headings in lesson content.

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+.
+
+Finished in 0.0032 seconds (files took 0.14864 seconds to load)
+1 example, 0 failures
+```
+
+At this point, refactoring isn't necessary. The `#add` method is essentially a one line method. As you progress in your Ruby learning, however, you might find your methods getting more complex, and you might find that you have to make extra efforts to abide by [SOLID](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)) principles. When that time comes, using RSpec and the "red-green-refactor" cycle will allow you to code with confidence, knowing that your classes and their behaviors continue to meet your specified expectations.
+
+## Assignments
+It's time to put your newfound knowledge to good use. Let's break our `Calculator` test.
```
Comment (KevinMulhern) [inline ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Can we wrap the assignment content in the assignment style div please?

```
<div class="lesson-content__panel" markdown="1">
  ...
</div>

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 10 above)
Comment (KevinMulhern) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Nice! 😄

File: ruby_programming/testing_with_rspec/test_driven_development.md [hunk 14]
```diff
@@ -0,0 +1,37 @@
+# Test-Driven Development
+
+## Introduction
```
Comment (KevinMulhern) [inline ruby_programming/testing_with_rspec/test_driven_development.md]: Can we convert the h2 headings in this file to h3's as well please?

File: ruby_programming/testing_with_rspec/test_driven_development.md [hunk 15]
```diff
@@ -0,0 +1,37 @@
+# Test-Driven Development
```
Comment (KevinMulhern) [inline ruby_programming/testing_with_rspec/test_driven_development.md]: Can we remove the heading on this file as well please?

File: ruby_programming/testing_with_rspec/test_driven_development.md [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+## Introduction
+
+Hopefully, you've been having fun developing in Ruby thus far. Perhaps one thing you may **not** have enjoyed as much, though, is having to run your script on the command line, running through each step manually, only to find that it doesn't return what you expected. And then comes the *even-less-fun* part: You tug at your hair a bit, change a few variables, and run your script again through each step, **hoping** that the results are correct.
+
+Folks, we're here to tell you there *IS* better. Hoping is for that exam you didn't study for and took by the seat of your pants. **Knowing** is test-driven development.
+
+## Learning Outcomes
+
+- Understand what test-driven development (TDD) is, and why it's important.
```
Comment (KevinMulhern) [inline ruby_programming/testing_with_rspec/test_driven_development.md]: Would adding "What are the three stages of a TDD cycle?" be a good additional learning outcome?

File: ruby_programming/testing_with_rspec/test_driven_development.md [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+
+Once the code is written and passes our test suite, you can move onto refactoring. As you might have guessed, running the test suite against the changes we make to our features or units enables us to refactor *while* making sure everything still works.
+
+Colloquially, this process is often referred to as the "**red-green-refactor**" cycle. That's all it is, in a nutshell: automated tests **drive** the design of software. You don't need to know how the entire architecture of your sweet, new, industry-breaking application will work. Your application only has to be broken down, step-by-step, until small units are identified and covered by tests.
+
+## Why Is It Important?
+
+Even if TDD intuitively seems like good practice, its utility is hotly debated even today. 
+
+No one is saying that tests aren't useful. Testing at [all levels](https://robots.thoughtbot.com/rails-test-types-and-the-testing-pyramid) is crucial. Testing ensures that your application is working as intended. Testing provides a safety net that gives your application room to grow and evolve as painlessly as possible as features are added and dependencies change.
```
Comment (KevinMulhern) [inline ruby_programming/testing_with_rspec/test_driven_development.md]: What do you think about changing this to:
"One of the less obvious but most important reasons for testing is the increase in confidence it gives developers when changing existing code.

Software backed by tests is a lot more open to change than software that isn't. Testing therefore provides a safety net that gives your application room to grow and evolve as painlessly as possible as features are added and dependencies change."

I'd like to emphasise the factor testing has in giving confidence to making changes in applications. But If you don't think its better the way it is I won't argue 😛

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 6 above)
Comment (CouchofTomato) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: Is rspec-rails a part of the standard rspec gem?

**Comment by KevinMulhern** [comment]
Left some minor comments, superb work 👏 💯

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 10 above)
Comment (CouchofTomato) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: haha. I concur on the shameless plug :)

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 6 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: @CouchofTomato it isn't. I don't see where if/where I mentioned that, but I'll look it over real quick.

File: ruby_programming/testing_with_rspec/test_driven_development.md (hunk 1 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/test_driven_development.md]: Good catch @105ron :star:

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 5 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: When making a concrete observation at a point in time that is likely to change in the future, I think it's proper to allude to the time of writing.

Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: @CouchofTomato Changed to "It's" in the second sentence, to maintain flow with the subordinate clauses.

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 6 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: @CouchofTomato found it :star:. Any thoughts on having the gems listed as bullets, rather than a phrase?

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 7 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: :+1:

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 9 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: :+1:

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 8 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: :+1:

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 10 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: :joy: :+1:

File: ruby_programming/testing_with_rspec/introduction_to_rspec.md (hunk 11 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/introduction_to_rspec.md]: :+1:

File: ruby_programming/testing_with_rspec/test_driven_development.md (hunk 16 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/test_driven_development.md]: Excellent idea :+1:

File: ruby_programming/testing_with_rspec/test_driven_development.md (hunk 17 above)
Comment (leosoaivan) [reply ruby_programming/testing_with_rspec/test_driven_development.md]: Yeah, excellent point. I tried to weave it in there, let me know if it gets too wordy.

**Comment by KevinMulhern** [comment]
👍 LGTM

@TheOdinProject/odinmods Is this ready to go live?

**Comment by mindovermiles262** [comment]
+1 from me
